Enable generating generic c_shim that doesn't bypass dispatcher

mikaylagawarecki opened the pull request:
Adds `c_shim_aten.{h/cpp}` and use this for `fill_`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158974

This is the generated `c_shim_aten.cpp` for reference

```cpp


// WARNING: THIS FILE IS AUTOGENERATED BY torchgen. DO NOT MODIFY BY HAND.
// See https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen.py#L2424-L2436 for details

// This file corresponds to the aten_shimified_ops list in torchgen/aoti/fallback_ops.py


#include <torch/csrc/inductor/aoti_torch/generated/c_shim_aten.h>
#include <torch/csrc/inductor/aoti_torch/utils.h>

#ifndef AT_PER_OPERATOR_HEADERS
#include <ATen/Functions.h>
#include <ATen/CompositeExplicitAutogradFunctions.h>
#include <ATen/CompositeExplicitAutogradNonFunctionalFunctions.h>
#include <ATen/CompositeImplicitAutogradFunctions.h>
#else
#include <ATen/ops/fill.h>

#endif // AT_PER_OPERATOR_HEADERS

using namespace torch::aot_inductor;


AOTITorchError aoti_torch_aten_fill__Scalar(AtenTensorHandle self, double value) {
    AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
        at::fill_(
            *tensor_handle_to_tensor_pointer(self), value
        );
```

mikaylagawarecki committed:
Enable generating generic c_shim that doesn't bypass dispatcher

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Enable generating generic c_shim that doesn't bypass dispatcher"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Enable generating generic c_shim that doesn't bypass dispatcher"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Enable generating generic c_shim that doesn't bypass dispatcher"

And use this for `fill_`


This is the generated `c_shim_generic.cpp` for reference

```cpp


// WARNING: THIS FILE IS AUTOGENERATED BY torchgen. DO NOT MODIFY BY HAND.
// See https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen.py#L2424-L2436 for details

#include <torch/csrc/inductor/aoti_torch/generated/c_shim_generic.h>
#include <torch/csrc/inductor/aoti_torch/utils.h>

#ifndef AT_PER_OPERATOR_HEADERS
#include <ATen/Functions.h>
#include <ATen/CompositeExplicitAutogradFunctions.h>
#include <ATen/CompositeExplicitAutogradNonFunctionalFunctions.h>
#include <ATen/CompositeImplicitAutogradFunctions.h>
#else
#include <ATen/ops/fill.h>

#endif // AT_PER_OPERATOR_HEADERS

using namespace torch::aot_inductor;


AOTITorchError aoti_torch_generic_fill__Scalar(AtenTensorHandle self, double value) {
    AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
        at::fill_(
            *tensor_handle_to_tensor_pointer(self), value
        );

    });
}
```



[ghstack-poisoned]

janeyx99 commented on caffe2/CMakeLists.txt:
```diff
@@ -267,6 +267,7 @@ if(NOT INTERN_DISABLE_AUTOGRAD AND NOT BUILD_LITE_INTERPRETER)
     "${TORCH_SRC_DIR}/csrc/autograd/generated/ADInplaceOrViewType_0.cpp"
     "${TORCH_SRC_DIR}/csrc/autograd/generated/ADInplaceOrViewType_1.cpp"
     "${TORCH_SRC_DIR}/csrc/inductor/aoti_torch/generated/c_shim_cpu.cpp"
+    "${TORCH_SRC_DIR}/csrc/inductor/aoti_torch/generated/c_shim_generic.cpp"
```
bikeshedding, might be clearer to call it c_shim_aten vs generic, idk

janeyx99 replied:
bikeshedding, might be clearer to call it c_shim_aten vs generic, idk

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -278,14 +278,29 @@ void boxed_empty_like(StableIValue* stack, uint64_t num_args, uint64_t num_outpu
   stack[0] = from(res);
 }
 
+Tensor my_fill_(Tensor t) {
```
```suggestion
Tensor fill_infinity(Tensor t) {
```

nit: more indicative

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/ops.py:
```diff
@@ -152,3 +152,15 @@ def my_zero_(t) -> Tensor:
     return torch.ops.libtorch_agnostic.my_zero_.default(t)
+
+
+def my_fill_(t) -> Tensor:
```
same here

janeyx99 commented on torch/csrc/inductor/aoti_torch/generated/c_shim_generic.h:
```diff
@@ -0,0 +1,18 @@
+
```
why are there two extra blank lines haha

mikaylagawarecki replied:
This isn't from my PR, all the rest of the c_shims are like this, so I'm leaving it as is

janeyx99 commented on torch/csrc/stable/ops.h:
```diff
@@ -24,6 +26,17 @@ inline Tensor empty_like(const Tensor& self) {
+// header-only.
+inline Tensor fill_(const Tensor& self, double value) {
+  auto self_ = self.get();
+  AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_generic_fill__Scalar(self_, value));
```
how come this cannot be
```suggestion
  AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_generic_fill__Scalar(self.get(), value));
```

janeyx99 commented on torch/csrc/inductor/aoti_torch/generated/c_shim_generic.h:
```diff
@@ -0,0 +1,18 @@
+
+
+// WARNING: THIS FILE IS AUTOGENERATED BY torchgen. DO NOT MODIFY BY HAND.
+// See https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen.py#L2424-L2436 for details
```
Might be good to specify that this file specifically corresponds to the generic dict

janeyx99 commented on torchgen/aoti/fallback_ops.py:
```diff
@@ -175,3 +175,12 @@
+# Operators in this list are intended to be used in torch/csrc/stable/ops.h
+# Unlike other c_shims, operators in this file do not bypass the dispatcher.
+# The same BC rules apply as inductor_fallback_ops.
+generic_fallback_ops: dict[str, dict[str, list[str]]] = {
```
Hehe so we had been hopping on the inductor fallback train, even though we were never using these as fallbacks. Since we're now diverging, i think it is better to name this to just 
```suggestion
aten_shimified_ops: dict[str, dict[str, list[str]]] = {
```

janeyx99 replied:
cc @desertfire just FYI, we are extending the functionality to codegen generic aten shims here!

janeyx99 commented on torchgen/gen_aoti_c_shim.py:
```diff
@@ -391,11 +392,14 @@ def gen_static_dispatch_backend_call_signature(
 def gen_static_dispatch_backend_call(
     f: NativeFunction,
-    backend_index: BackendIndex,
+    backend_index: Optional[BackendIndex] = None,
```
i think this is fine yea

janeyx99 commented on torchgen/gen_aoti_c_shim.py:
```diff
@@ -391,11 +392,14 @@ def gen_static_dispatch_backend_call_signature(
+    if backend_index is None:
+        return f"at::{cpp_sig.name()}"
+    else:
+        return f"at::{backend_index.dispatch_key.lower()}::{cpp_sig.name()}"
```
this code change looks good to me (and works). cc @albanD--I'd appreciate if you took a glance at this codegen portion and ascertain that we're making the right assumptions!

albanD reviewed (approved):
No concern beyond the fake dispatch key on my end!

mikaylagawarecki committed:
Update on "Enable generating generic c_shim that doesn't bypass dispatcher"

And use this for `fill_`


This is the generated `c_shim_generic.cpp` for reference

```cpp


// WARNING: THIS FILE IS AUTOGENERATED BY torchgen. DO NOT MODIFY BY HAND.
// See https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen.py#L2424-L2436 for details

#include <torch/csrc/inductor/aoti_torch/generated/c_shim_generic.h>
#include <torch/csrc/inductor/aoti_torch/utils.h>

#ifndef AT_PER_OPERATOR_HEADERS
#include <ATen/Functions.h>
#include <ATen/CompositeExplicitAutogradFunctions.h>
#include <ATen/CompositeExplicitAutogradNonFunctionalFunctions.h>
#include <ATen/CompositeImplicitAutogradFunctions.h>
#else
#include <ATen/ops/fill.h>

#endif // AT_PER_OPERATOR_HEADERS

using namespace torch::aot_inductor;


AOTITorchError aoti_torch_generic_fill__Scalar(AtenTensorHandle self, double value) {
    AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
        at::fill_(
            *tensor_handle_to_tensor_pointer(self), value
        );

    });
}
```



[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Enable generating generic c_shim that doesn't bypass dispatcher"

And use this for `fill_`


This is the generated `c_shim_generic.cpp` for reference

```cpp


// WARNING: THIS FILE IS AUTOGENERATED BY torchgen. DO NOT MODIFY BY HAND.
// See https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen.py#L2424-L2436 for details

#include <torch/csrc/inductor/aoti_torch/generated/c_shim_generic.h>
#include <torch/csrc/inductor/aoti_torch/utils.h>

#ifndef AT_PER_OPERATOR_HEADERS
#include <ATen/Functions.h>
#include <ATen/CompositeExplicitAutogradFunctions.h>
#include <ATen/CompositeExplicitAutogradNonFunctionalFunctions.h>
#include <ATen/CompositeImplicitAutogradFunctions.h>
#else
#include <ATen/ops/fill.h>

#endif // AT_PER_OPERATOR_HEADERS

using namespace torch::aot_inductor;


AOTITorchError aoti_torch_generic_fill__Scalar(AtenTensorHandle self, double value) {
    AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
        at::fill_(
            *tensor_handle_to_tensor_pointer(self), value
        );

    });
}
```



[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Enable generating generic c_shim that doesn't bypass dispatcher"

And use this for `fill_`


This is the generated `c_shim_generic.cpp` for reference

```cpp


// WARNING: THIS FILE IS AUTOGENERATED BY torchgen. DO NOT MODIFY BY HAND.
// See https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen.py#L2424-L2436 for details

#include <torch/csrc/inductor/aoti_torch/generated/c_shim_generic.h>
#include <torch/csrc/inductor/aoti_torch/utils.h>

#ifndef AT_PER_OPERATOR_HEADERS
#include <ATen/Functions.h>
#include <ATen/CompositeExplicitAutogradFunctions.h>
#include <ATen/CompositeExplicitAutogradNonFunctionalFunctions.h>
#include <ATen/CompositeImplicitAutogradFunctions.h>
#else
#include <ATen/ops/fill.h>

#endif // AT_PER_OPERATOR_HEADERS

using namespace torch::aot_inductor;


AOTITorchError aoti_torch_generic_fill__Scalar(AtenTensorHandle self, double value) {
    AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
        at::fill_(
            *tensor_handle_to_tensor_pointer(self), value
        );

    });
}
```



[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Enable generating generic c_shim that doesn't bypass dispatcher"

And use this for `fill_`


This is the generated `c_shim_aten.cpp` for reference

```cpp


// WARNING: THIS FILE IS AUTOGENERATED BY torchgen. DO NOT MODIFY BY HAND.
// See https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen.py#L2424-L2436 for details

// This file corresponds to the aten_shimified_ops list in torchgen/aoti/fallback_ops.py


#include <torch/csrc/inductor/aoti_torch/generated/c_shim_aten.h>
#include <torch/csrc/inductor/aoti_torch/utils.h>

#ifndef AT_PER_OPERATOR_HEADERS
#include <ATen/Functions.h>
#include <ATen/CompositeExplicitAutogradFunctions.h>
#include <ATen/CompositeExplicitAutogradNonFunctionalFunctions.h>
#include <ATen/CompositeImplicitAutogradFunctions.h>
#else
#include <ATen/ops/fill.h>

#endif // AT_PER_OPERATOR_HEADERS

using namespace torch::aot_inductor;


AOTITorchError aoti_torch_aten_fill__Scalar(AtenTensorHandle self, double value) {
    AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
        at::fill_(
            *tensor_handle_to_tensor_pointer(self), value
        );
```



[ghstack-poisoned]

mikaylagawarecki commented on torchgen/model.py:
```diff
@@ -229,6 +229,10 @@ class DispatchKey(Enum):
 
+    # GenericKey is not a real DispatchKey
+    # It is used for generating generic C shim headers for stable operators
+    GenericKey = auto()
```
Fixed

mikaylagawarecki committed:
Update on "Enable generating generic c_shim that doesn't bypass dispatcher"

And use this for `fill_`


This is the generated `c_shim_aten.cpp` for reference

```cpp


// WARNING: THIS FILE IS AUTOGENERATED BY torchgen. DO NOT MODIFY BY HAND.
// See https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen.py#L2424-L2436 for details

// This file corresponds to the aten_shimified_ops list in torchgen/aoti/fallback_ops.py


#include <torch/csrc/inductor/aoti_torch/generated/c_shim_aten.h>
#include <torch/csrc/inductor/aoti_torch/utils.h>

#ifndef AT_PER_OPERATOR_HEADERS
#include <ATen/Functions.h>
#include <ATen/CompositeExplicitAutogradFunctions.h>
#include <ATen/CompositeExplicitAutogradNonFunctionalFunctions.h>
#include <ATen/CompositeImplicitAutogradFunctions.h>
#else
#include <ATen/ops/fill.h>

#endif // AT_PER_OPERATOR_HEADERS

using namespace torch::aot_inductor;


AOTITorchError aoti_torch_aten_fill__Scalar(AtenTensorHandle self, double value) {
    AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
        at::fill_(
            *tensor_handle_to_tensor_pointer(self), value
        );
```



[ghstack-poisoned]

janeyx99 reviewed (approved):
cool thanks!

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki commented:
@pytorchbot merge
